processing/src/main/java/io/druid/segment/FloatDimensionIndexer.java

```diff
@@ -191,21 +193,51 @@ public Object get()
   }
 
   @Override
-  public int compareUnsortedEncodedKeyComponents(Float lhs, Float rhs)
+  public DoubleColumnSelector makeDoubleColumnSelector(
+      IncrementalIndexStorageAdapter.EntryHolder currEntry, IncrementalIndex.DimensionDesc desc
+  )
+  {
+    final int dimIndex = desc.getIndex();
+    class IndexerDoubleColumnSelector implements DoubleColumnSelector
+    {
+      @Override
+      public double get()
+      {
+        final Object[] dims = currEntry.getKey().getDims();
+
+        if (dimIndex >= dims.length) {
+          return 0.0;
+        }
+        float floatVal = (Float) dims[dimIndex];
+        return (double) floatVal;
+      }
+
+      @Override
+      public void inspectRuntimeShape(RuntimeShapeInspector inspector)
+      {
+        // nothing to inspect
+      }
+    }
+
+    return new IndexerDoubleColumnSelector();
+  }
+
+  @Override
+  public int compareUnsortedEncodedKeyComponents(@Nullable Float lhs, @Nullable Float rhs)
   {
-    return lhs.compareTo(rhs);
+    return DimensionHandlerUtils.nullToZero(lhs).compareTo(DimensionHandlerUtils.nullToZero(rhs));
   }
 
   @Override
-  public boolean checkUnsortedEncodedKeyComponentsEqual(Float lhs, Float rhs)
+  public boolean checkUnsortedEncodedKeyComponentsEqual(@Nullable Float lhs, @Nullable Float rhs)
   {
-    return lhs.equals(rhs);
+    return Objects.equals(DimensionHandlerUtils.nullToZero(lhs), DimensionHandlerUtils.nullToZero(rhs));
   }
 
   @Override
-  public int getUnsortedEncodedKeyComponentHashCode(Float key)
+  public int getUnsortedEncodedKeyComponentHashCode(@Nullable Float key)
   {
-    return key.hashCode();
+    return Objects.hashCode(DimensionHandlerUtils.nullToZero(key));
```

Comment:
Similar to above, suggested `DimensionHandlerUtils.nullToZero(key).hashCode()`. + Same for Double, Long.